// _@reactivex_rxjs@6.4.0@@reactivex/rxjs/tmp/tests/rxjs/internal/create/fromEvent-spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var testing_1 = require("rxjs/testing");
var assertDeepEquals_1 = require("../test_helpers/assertDeepEquals");
/** @test {fromEvent} */
describe('fromEvent', function () {
    var testScheduler;
    beforeEach(function () {
        testScheduler = new testing_1.TestScheduler(assertDeepEquals_1.assertDeepEquals);
    });
    //asDiagram('fromEvent(element, \'click\')')
    it('should create an observable of click on the element', function () {
        testScheduler.run(function (_a) {
            var hot = _a.hot, cold = _a.cold, expectObservable = _a.expectObservable, expectSubscriptionsTo = _a.expectSubscriptionsTo;
            var target = {
                addEventListener: function (eventType, listener) {
                    rxjs_1.timer(5, 2, testScheduler)
                        .pipe(operators_1.mapTo('ev'), operators_1.take(2), operators_1.concatWith(rxjs_1.NEVER))
                        .subscribe(function (e) { return listener(e); });
                },
                removeEventListener: function () { return void 0; },
                dispatchEvent: function () { return void 0; },
            };
            var e1 = rxjs_1.fromEvent(target, 'click');
            var expected = '-----x-x---';
            expectObservable(e1).toBe(expected, { x: 'ev' });
        });
    });
    it('should setup an event observable on objects with "on" and "off" ', function () {
        var onEventName;
        var onHandler;
        var offEventName;
        var offHandler;
        var obj = {
            on: function (a, b) {
                onEventName = a;
                onHandler = b;
            },
            off: function (a, b) {
                offEventName = a;
                offHandler = b;
            }
        };
        var subscription = rxjs_1.fromEvent(obj, 'click')
            .subscribe(function () {
            //noop
        });
        subscription.unsubscribe();
        chai_1.expect(onEventName).to.equal('click');
        chai_1.expect(typeof onHandler).to.equal('function');
        chai_1.expect(offEventName).to.equal(onEventName);
        chai_1.expect(offHandler).to.equal(onHandler);
    });
    it('should setup an event observable on objects with "addEventListener" and "removeEventListener" ', function () {
        var onEventName;
        var onHandler;
        var offEventName;
        var offHandler;
        var obj = {
            addEventListener: function (a, b, useCapture) {
                onEventName = a;
                onHandler = b;
            },
            removeEventListener: function (a, b, useCapture) {
                offEventName = a;
                offHandler = b;
            }
        };
        var subscription = rxjs_1.fromEvent(obj, 'click')
            .subscribe(function () {
            //noop
        });
        subscription.unsubscribe();
        chai_1.expect(onEventName).to.equal('click');
        chai_1.expect(typeof onHandler).to.equal('function');
        chai_1.expect(offEventName).to.equal(onEventName);
        chai_1.expect(offHandler).to.equal(onHandler);
    });
    it('should setup an event observable on objects with "addListener" and "removeListener" returning event emitter', function () {
        var onEventName;
        var onHandler;
        var offEventName;
        var offHandler;
        var obj = {
            addListener: function (a, b) {
                onEventName = a;
                onHandler = b;
                return this;
            },
            removeListener: function (a, b) {
                offEventName = a;
                offHandler = b;
                return this;
            }
        };
        var subscription = rxjs_1.fromEvent(obj, 'click')
            .subscribe(function () {
            //noop
        });
        subscription.unsubscribe();
        chai_1.expect(onEventName).to.equal('click');
        chai_1.expect(typeof onHandler).to.equal('function');
        chai_1.expect(offEventName).to.equal(onEventName);
        chai_1.expect(offHandler).to.equal(onHandler);
    });
    it('should setup an event observable on objects with "addListener" and "removeListener" returning nothing', function () {
        var onEventName;
        var onHandler;
        var offEventName;
        var offHandler;
        var obj = {
            addListener: function (a, b, context) {
                onEventName = a;
                onHandler = b;
                return { context: '' };
            },
            removeListener: function (a, b) {
                offEventName = a;
                offHandler = b;
            }
        };
        var subscription = rxjs_1.fromEvent(obj, 'click')
            .subscribe(function () {
            //noop
        });
        subscription.unsubscribe();
        chai_1.expect(onEventName).to.equal('click');
        chai_1.expect(typeof onHandler).to.equal('function');
        chai_1.expect(offEventName).to.equal(onEventName);
        chai_1.expect(offHandler).to.equal(onHandler);
    });
    it('should setup an event observable on objects with "addListener" and "removeListener" and "length" ', function () {
        var onEventName;
        var onHandler;
        var offEventName;
        var offHandler;
        var obj = {
            addListener: function (a, b) {
                onEventName = a;
                onHandler = b;
            },
            removeListener: function (a, b) {
                offEventName = a;
                offHandler = b;
            },
            length: 1
        };
        var subscription = rxjs_1.fromEvent(obj, 'click')
            .subscribe(function () {
            //noop
        });
        subscription.unsubscribe();
        chai_1.expect(onEventName).to.equal('click');
        chai_1.expect(typeof onHandler).to.equal('function');
        chai_1.expect(offEventName).to.equal(onEventName);
        chai_1.expect(offHandler).to.equal(onHandler);
    });
    it('should error on invalid event targets', function () {
        var obj = {
            addListener: function () {
                //noop
            }
        };
        rxjs_1.fromEvent(obj, 'click').subscribe({
            error: function (err) {
                chai_1.expect(err).to.exist
                    .and.be.instanceof(Error)
                    .and.have.property('message', 'Invalid event target');
            }
        });
    });
    it('should pass through options to addEventListener and removeEventListener', function () {
        var onOptions;
        var offOptions;
        var expectedOptions = { capture: true, passive: true };
        var obj = {
            addEventListener: function (a, b, c) {
                onOptions = c;
            },
            removeEventListener: function (a, b, c) {
                offOptions = c;
            }
        };
        var subscription = rxjs_1.fromEvent(obj, 'click', expectedOptions)
            .subscribe(function () {
            //noop
        });
        subscription.unsubscribe();
        chai_1.expect(onOptions).to.equal(expectedOptions);
        chai_1.expect(offOptions).to.equal(expectedOptions);
    });
    it('should pass through events that occur', function (done) {
        var send;
        var obj = {
            on: function (name, handler) {
                send = handler;
            },
            off: function () {
                //noop
            }
        };
        rxjs_1.fromEvent(obj, 'click').pipe(operators_1.take(1))
            .subscribe({
            next: function (e) {
                chai_1.expect(e).to.equal('test');
            },
            error: function () {
                done(new Error('should not be called'));
            },
            complete: function () {
                done();
            }
        });
        send('test');
    });
    it('should emit multiple arguments from event as an array', function (done) {
        var send;
        var obj = {
            on: function (name, handler) {
                send = handler;
            },
            off: function () {
                //noop
            }
        };
        rxjs_1.fromEvent(obj, 'click').pipe(operators_1.take(1))
            .subscribe({
            next: function (e) {
                chai_1.expect(e).to.deep.equal([1, 2, 3]);
            },
            error: function () {
                done(new Error('should not be called'));
            },
            complete: function () {
                done();
            }
        });
        send(1, 2, 3);
    });
    it('should not throw an exception calling toString on obj with a null prototype', function (done) {
        // NOTE: Can not test with Object.create(null) or `class Foo extends null`
        // due to TypeScript bug. https://github.com/Microsoft/TypeScript/issues/1108
        var NullProtoEventTarget = /** @class */ (function () {
            function NullProtoEventTarget() {
            }
            NullProtoEventTarget.prototype.on = function () { };
            NullProtoEventTarget.prototype.off = function () { };
            return NullProtoEventTarget;
        }());
        NullProtoEventTarget.prototype.toString = null;
        var obj = new NullProtoEventTarget();
        chai_1.expect(function () {
            rxjs_1.fromEvent(obj, 'foo').subscribe();
            done();
        }).to.not.throw(TypeError);
    });
});
